ambry-commons/src/main/java/com/github/ambry/commons/HardwareUsageMeter.java

```diff
@@ -0,0 +1,84 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.commons;
+
+import java.lang.management.BufferPoolMXBean;
+import java.lang.management.ManagementFactory;
+import com.sun.management.OperatingSystemMXBean;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A class to report hardware usage.
+ */
+public class HardwareUsageMeter {
+  private static final Logger logger = LoggerFactory.getLogger(HardwareUsageMeter.class);
+  private AtomicLong cpuLastSampleTime = new AtomicLong(0);
+  private final int cpuSamplingPeriodMs;
+  private int cpuPercentage;
+  private AtomicLong memoryLastSampleTime = new AtomicLong(0);
+  private final int memorySamplingPeriodMs;
+  private int memoryPercentage;
+  private long maxMemory = Runtime.getRuntime().maxMemory();
+  private BufferPoolMXBean directMemoryMxBean;
+  private OperatingSystemMXBean osBean;
+
+  public HardwareUsageMeter(int cpuSamplingPeriodMs, int memorySamplingPeriodMs) {
+    this.cpuSamplingPeriodMs = cpuSamplingPeriodMs;
+    this.memorySamplingPeriodMs = memorySamplingPeriodMs;
+
+    List<BufferPoolMXBean> pools = ManagementFactory.getPlatformMXBeans(BufferPoolMXBean.class);
+    directMemoryMxBean = pools.stream().filter(bean -> bean.getName().equals("direct")).findFirst().orElse(null);
+    if (directMemoryMxBean == null) {
+      logger.error("Couldn't get directMemoryMxBean");
+    }
+    osBean = ManagementFactory.getPlatformMXBean(OperatingSystemMXBean.class);
+  }
+
+  private int getCpuPercentage() {
+    // In experiments, osBean.getSystemCpuLoad() takes 15-25 ms
+    if (System.currentTimeMillis() <= cpuLastSampleTime.get() + cpuSamplingPeriodMs) {
+      return cpuPercentage;
+    }
+    cpuPercentage = (int) (osBean.getSystemCpuLoad() * 100);
+    cpuLastSampleTime.set(System.currentTimeMillis());
+    logger.trace("CPU percentage: {}", cpuPercentage);
+    return cpuPercentage;
+  }
+
+  private int getMemoryPercentage() {
+    // In experiments, Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory() takes 2-6ms
+    // directMemoryMxBean.getMemoryUsed() takes 2-15ms
+    if (System.currentTimeMillis() <= memoryLastSampleTime.get() + memorySamplingPeriodMs) {
+      return memoryPercentage;
+    }
+    long usedHeapMemory = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+    memoryPercentage = (int) ((usedHeapMemory + directMemoryMxBean.getMemoryUsed()) * 100 / maxMemory);
```

Comment:
Just a clarifying question: is maxMemory the sum of max heap size and max direct memory size?

Reply:
Hi Casey, 

The `maxMemory` is close to `-Xmx`, based on my test(1%-5% less).
I assumed it's heap+direct memory. Because in our previous heap OOM, where 32GB Xmx was set, most of heap OOM's heap dump were much smaller than 32GB(12GB or 20GB). So I thought `Xms` includes heap + direct, then `maxMemory` should be the max for heap + direct.

However, I double checked `Xms` option, it's heap only. [This answer](https://stackoverflow.com/questions/13988328/java-memory-runtime-getruntime-maxmemory) also states it doesn't include direct memory. 

I am confused now, if Xms is heap only, why our heap dumps are much less than 32GB? 

----
For this PR, I can revise the code to return heap usage only in `getMemoryPercentage`(will be renamed to getHeapMemoryPercentage), and add `getDirectMemoryPercentage` in next PR(haven't got a good way to get max direct memory.)

Reply:
I'm not sure why our heap dumps are smaller. It could be the difference between a full heap dump that also includes yet-to-be garbage collected objects and a regular heap dump.

Reply:
Found this: https://stackoverflow.com/questions/3773775/default-for-xxmaxdirectmemorysize

looks like our servers doesn't specify `xxmaxdirectmemorysize` so total allocation for heap+direct is `2*Runtime.getRuntime().maxMemory()`

Reply:
I wonder if we could either read the property `xxmaxdirectmemorysize` or, if it is not set, use max heap size.

We do set this property on ambry-server to support in memory indexes, but likely not on ambry-frontend.

Reply:
We could. 

`sun.misc.VM.maxDirectMemory()`

If `xxmaxdirectmemorysize` is not set, it returns the same value of `Runtime.getRuntime().maxMemory()`. If set, it returns the value of `xxmaxdirectmemorysize`. 

Tested in my local, returned expected value.

Reply:
The only problem with sun packages is that they are private-ish (SDK internal) APIs, and there are problems accessing them in java 11: https://stackoverflow.com/questions/53543062/replace-access-to-sun-misc-vm-for-jdk-11

Reply:
> @zzmao since our travis build is on java 11, it will fail when using sun.misc packages:
> 
> ```
> /home/travis/build/linkedin/ambry/ambry-commons/src/main/java/com/github/ambry/commons/HardwareUsageMeter.java:40: error: package sun.misc does not exist
>   private long maxDirectMemory = sun.misc.VM.maxDirectMemory();
>                                          ^
> Note: /home/travis/build/linkedin/ambry/ambry-commons/src/main/java/com/github/ambry/commons/CommonUtils.java uses or overrides a deprecated API.
> ```

Just saw your previous comment. 

We could also parse the vm option:

```
HotSpotDiagnosticMXBean hsdiag = ManagementFactory
     .getPlatformMXBean(HotSpotDiagnosticMXBean.class);
hsdiag.getVMOption("MaxDirectMemorySize")
```

Reply:
Yeah maybe that's the best approach. If it is set, use the value of the vm option, if not set, use the heap size.